src/dataMap/metaManager/lazyFactoryMap.js

```diff
@@ -0,0 +1,265 @@
+ * where the new rows are added are upshifted by 2.
+ *               │
+ *               │ Insert 2 rows
+ *              \│/
+ * +------+------+------+------+------+
+ * | 0/10 | 1/11 | 2/12 | 3/13 | 4/14 |  Keys before
+ * +------+------+------+------+------+
+ *
+ *                / 2 new rows \
+ * +------+------+------+------+------+------+------+
+ * | 0/10 | 1/11 | 2/12 | 3/13 | 4/14 | 5/15 | 6/16 |  Keys after
+ * +------+------+------+------+------+------+------+
+ *    │       │      │      │      │      │     │
+ *    │       │      └──────┼──────┼──────┼┐    │
+ *    │       │             └──────┼──────┼┼────┼┐
+ *    │       │      ┌─────────────┘      ││    ││
+ *    │       │      │      ┌─────────────┘│    ││
+ *    │       │      │      │      ┌───────┼────┘│
+ *    │       │      │      │      │       │     │
+ * +------+------+------+------+------+------+------+
+ * | AAA  | BBB  | CCC  | DDD  | EEE  | FFF  | GGG  |  Data
+ * +------+------+------+------+------+------+------+
+ *
+ * Now at index 2 and 3 we have access to new items.
+ *
+ * map.obtain(2) // returns new value "FFF" for newly created row.
+ * map.obtain(4) // index shifted by 2 has access to the old "CCC" value, as before inserting.
+ *
+ * after removing 2 rows, keys that hold the data positioned after the place where the
+ * rows are removed are downshifted by 2.
+ *               │
+ *               │ Remove 2 rows
+ *               ├─────────────┐
+ *              \│/            │
+ * +------+------+------+------+------+
+ * | 0/10 | 1/11 | 2/12 | 3/13 | 4/14 |  Keys before
+ * +------+------+------+------+------+
+ *
+ * +------+------+------+
+ * | 0/10 | 1/11 | 2/14 |  Keys after
+ * +------+------+------+
+ *    │       │      │
+ *    │       │      └─────────────┐
+ *    │       │                    │
+ *    │       │                    │
+ *    │       │                    │
+ *    │       │                    │
+ *    │       │                    │
+ * +------+------+------+------+------+
+ * | AAA  | BBB  | CCC  | DDD  | EEE  |  Data
+ * +------+------+------+------+------+
+ *                  /│\   /|\
+ *                   └──┬──┘
+ *         This data is marked as "hole" which
+ *         means that can be replaced by new item
+ *         when that will be created.
+ *
+ * map.obtain(2) // returns the old value ("EEE") as it should
+ */
+export default class LazyFactoryMap {
+  constructor(valueFactory) {
+    this.valueFactory = valueFactory;
+    /**
+     * An array which contains data.
+     *
+     * @type {Array}
+     */
+    this.data = [];
+    /**
+     * An array of indexes where the key of the array is mapped to the value which points to the
+     * specific position of the data array.
+     *
+     * @type {Number[]}
+     */
+    this.index = [];
+    /**
+     * An array of indexes where points to the data items which can be replaced by obtaining new
```

Comment:
For me, the `side effect` isn't the best phrase describing how it works (if I understand it properly). This map has no side effect i.e. for memory. You leave some data with premeditation for further work. I would change that description, my proposition is: `The "holes" are an intended effect of deleting entries.`. For me, this description isn’t clear yet.

Information about purpose, why you leave some data in the memory would help a lot.

Reply:
I've added a little bit more description to this https://github.com/handsontable/handsontable/pull/6254/commits/eb1f3258945f22e65a1255460a62cc3da680e58e#diff-6401234415dbf15f9803715a2e1becc1R107